waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -507,10 +535,40 @@
           (<= attempt max-patch-retries)
           (recur (inc attempt) (get-replica-count scheduler service-id)))))))
 
-(defn- kill-service-instance
+(let [num-deletion-locks 100
+      deletion-locks (repeatedly num-deletion-locks (fn [] (Object.)))]
+  (defn- service-id->deletion-lock
+    "Returns the deletion lock striped by a hash of the service ID."
+    [service-id]
+    (nth deletion-locks (mod (hash service-id) num-deletion-locks))))
+
+(defn- instance->pod-url
+  "Returns the pod api server url."
+  [api-server-url {:keys [k8s/namespace k8s/pod-name]}]
+  (str api-server-url "/api/v1/namespaces/" namespace "/pods/" pod-name))
+
+(defn hard-delete-service-instance
+  "Force kill the Kubernetes pod corresponding to the given Waiter Service Instance.
+   Does not adjust ReplicaSet replica count; preventing scheduling of a replacement pod must be ensured by the callee.
+   Uses locking to prevent concurrent calls for killing instances from the same service.
+   Returns nil on success, but throws on failure."
+  [{:keys [api-server-url] :as scheduler} {:keys [service-id] :as instance}]
+  (locking
+    (service-id->deletion-lock service-id)
+    (let [pod-url (instance->pod-url api-server-url instance)]
+      (try
+        ; "hard" delete the pod (i.e., actually kill, allowing the pod's default grace period expires)
+        ; (note that the pod's default grace period is different from the 300s period set above)
```

Comment:
This comment no longer makes sense because you moved this logic out of and _above_ the original function.